src/main/java/net/glowstone/entity/GlowLivingEntity.java

```diff
@@ -830,6 +849,47 @@ public void damage(double amount, Entity source, DamageCause cause) {
         setLastDamager(source);
     }
 
+    /**
+     * Checks if the incoming source of damage is a "Player Hit".
+     * @param source The incoming source of damage
+     */
+    private boolean isPlayerHit(Entity source) {
+        // If directly damaged by a player
+        if (source instanceof GlowPlayer) {
+            return true;
+        // If damaged by a TNT ignited by a player
+        } else if (source instanceof GlowTntPrimed) {
+            return
+                ((GlowTntPrimed)source).getPlayer() != null
+                && (((GlowTntPrimed)source).getPlayer().getGameMode() == GameMode.SURVIVAL
+                || ((GlowTntPrimed)source).getPlayer().getGameMode() == GameMode.ADVENTURE);
+        // If damaged by a tamed wolf
+        } else if (source instanceof GlowWolf) {
+            return ((GlowWolf)source).isTamed();
+        }
+        return false;
+    }
+
+    /**
+     * Determines the player who did the damage from source of damage.
+     * @param source The incoming source of damage
+     */
+    private Player determinePlayer(Entity source) {
+        // If been killed by an ignited tnt
+        if (source instanceof GlowTntPrimed) {
+            return (Player)((GlowTntPrimed)source).getPlayer();
+        }
+        // If been killed by a player
+        else if (source instanceof GlowPlayer) {
```

Comment:
Since you are returning inside the if-blocks, the `else`s are redundant and can be omitted.

Reply:
Should we maybe use a VicariousKillerEntity interface to simplify all of this?

Reply:
I don't think that's necessary. There are only several cases that require checking vicarious killing.